[SYCL] Add offload wrapping for SYCL kind.

maksimsab opened the pull request:
This patch adds an Offload Wrapper for the SYCL kind. This is an essential step for SYCL offloading and the compilation flow. The usage of offload wrapping is added to the clang-linker-wrapper tool.

Modifications:
Implemented `bundleSYCL()` function to handle SYCL image bundling.
Implemented `wrapSYCLBinaries()` function that is invoked from clang-linker-wrapper.
SYCL Offload Wrapping uses specific data structures such as `__sycl.tgt_device_image` and `__sycl.tgt_bin_desc`. Each SYCL image maintains its own symbol table (unlike shared global tables in other targets). Therefore, symbols are encoded explicitly during the offload wrapping. Also, images refer to their own Offloading Entries arrays unlike other targets.
The proposed `__sycl.tgt_device_image` uses Version 3 to differentiate from images generated by Intel DPC++. The structure proposed in this patch doesn't have fields deprecated in DPC++.

maksimsab committed:
[SYCL] Add offload wrapping for SYCL kind.

maksimsab commented:
Hi @jdoerfert @jhuber6 !

We are adding offload wrapping for SYCL in this PR. Do you have any suggestions for whom I should request for a code review?

jhuber6 reviewed (commented):
Why do we need a separate directory for the offload wrapper?

jhuber6 commented on clang/test/Driver/linker-wrapper-image.c:
```diff
@@ -263,3 +264,37 @@
+// SYCL-NEXT: @.sycl_offloading.0.info = internal local_unnamed_addr constant [2 x i64] [i64 ptrtoint (ptr @.sycl_offloading.0.data to i64), i64 0], section ".tgtimg", align 16
+// SYCL-NEXT: @llvm.used = appending global [1 x ptr] [ptr @.sycl_offloading.0.info], section "llvm.metadata"
+// SYCL-NEXT: @.sycl_offloading.device_images = internal unnamed_addr constant [1 x %__sycl.tgt_device_image] [%__sycl.tgt_device_image { i16 3, i8 8, i8 0, ptr @.sycl_offloading.target.0, ptr @.sycl_offloading.opts.compile.0, ptr @.sycl_offloading.opts.link.0, ptr @.sycl_offloading.0.data, ptr @.sycl_offloading.0.data, ptr null, ptr null, ptr null, ptr null }]
+// SYCL-NEXT: @.sycl_offloading.descriptor = internal constant %__sycl.tgt_bin_desc { i16 1, i16 1, ptr @.sycl_offloading.device_images, ptr null, ptr null }
```
Is this just copied from OpenMP?

maksimsab replied:
`RUN` lines are copied from OpenMP and adjusted to SYCL case. Output's checks are copied from the `clang-linker-wrapper`'s output.

maksimsab replied:
Initially, we customized old `clang-offload-wrapper` and then we extracted SYCL logic in a separate place. That is how these names stuck with us.

maksimsab committed:
Merge branch 'main' into public_offload_wrapper

maksimsab committed:
consolidate SYCL wrapper with others.

maksimsab commented:
I've merged SYCL wrapper with all others.

jhuber6 commented on llvm/lib/Frontend/CMakeLists.txt:
```diff
@@ -5,4 +5,4 @@ add_subdirectory(HLSL)
 add_subdirectory(OpenMP)
 add_subdirectory(Offloading)
-add_subdirectory(SYCL)
+#add_subdirectory(SYCL)
```
Should be deleted if it was removed.

maksimsab replied:
Sorry, I didn't notice that. Now that looks clean.

maksimsab committed:
Merge branch 'main' into public_offload_wrapper

jhuber6 commented on llvm/lib/Frontend/Offloading/OffloadWrapper.cpp:
```diff
@@ -620,6 +635,422 @@ void createRegisterFatbinFunction(Module &M, GlobalVariable *FatbinDesc,
+  ///
+  /// \returns Pair of Constants that point at entries content.
+  std::pair<Constant *, Constant *>
+  addOffloadEntriesToModule(StringRef Entries) {
```
Is this stuff not common with the other targets?

maksimsab replied:
Entries could be emitted by `emitOffloadEntry`. Then I should create an array that refers to these entries.
I don't see the appropriate other function that does that. The most closest one is `offloading::getOffloadEntryArray` but this function creates an empty array if I get it right. Therefore, I would need to modify that function in a strange way.
If you have other opinion regarding `getOffloadEntryArray` please let me know.

maksimsab replied:
That is confusing. `linker-wrapper-image.c` have zero entry points in the input and it definitely has checked zero initialized entry arrays. Therefore, the output of the mentioned linking mechanism is either tested somewhere else or isn't tested at all. If it isn't tested then I could try to add such testing.

maksimsab replied:
Let's consider the following example.
One TU has 2 kernels (A, B) and the user requests a split by kernels.
In unified approach we would have the output from the Codegen like the following:
```
kernel A() {...}
kernel B() {...}

@entry_A = %tgt_offload_entry {...} ; info about kernel A
@entry_B = %tgt_offload_entry {...} ; info about kernel B. 
```

Then we have to split these kernels using `splitModuleTransitiveFromEntryPoints` added in #131347. In order to correctly move these entries, we have to add such SYCL specific handling functionality in `splitModuleTransitiveFromEntryPoints`, which is very discouraged according to my understanding. That is only my concern about that.

YuriPlyakhin replied:
@jhuber6, we replaced duplicated code with usage of `emitOffloadingEntry` and updated name and description of `addOffloadEntriesToModule` (now it is `initOffloadEntriesPerImage`) to avoid confusion with `getOffloadEntryArray`, because initially similar names created perception that the functions are doing the same, but in reality, they are different.

We need to initialize offload entries per image, because, unlike other targets, each SYCL device image contains its own set of symbols, which may contain different symbols than other images. The reason is per-kernel device code split, which solves the following problems:
1. SYCL 2020 spec requirement: “An implementation may not raise a compile time diagnostic or a run time exception merely due to speculative compilation of a kernel for a device when the application does not actually submit the kernel to that device.” To satisfy this requirement, we split kernels into different images based on the features required.
2. Compilation overhead. There are use cases, when module contains 1000s of kernels. Providing only specific symbols in each of the images allows to compile/run 1 specific kernel/TU, so no need to JIT compile all 1000s.

Does that address all your concerns for this PR?

maksimsab committed:
Remove trash leftovers

YuriPlyakhin commented on llvm/include/llvm/Frontend/Offloading/OffloadWrapper.h:
```diff
@@ -52,6 +54,23 @@ LLVM_ABI llvm::Error wrapHIPBinary(llvm::Module &M, llvm::ArrayRef<char> Images,
+  // target/compiler specific options what are suggested to use to "compile"
+  // program at runtime.
+  std::string CompileOptions;
+  // Target/Compiler specific options that are suggested to use to "link"
```
nit
```suggestion
  // Target/compiler specific options that are suggested to use to "link"
```

maksimsab committed:
Merge branch 'main' into public_offload_wrapper

maksimsab committed:
address some CR items

YuriPlyakhin commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -717,6 +717,14 @@ wrapDeviceImages(ArrayRef<std::unique_ptr<MemoryBuffer>> Buffers,
+    offloading::SYCLWrappingOptions WrappingOptions;
+    if (Error Err =
+            offloading::wrapSYCLBinaries(M, BuffersToWrap, WrappingOptions))
+      return Err;
```
nit
```suggestion
      return std::move(Err);
```

maksimsab committed:
Reuse emitOffloadingEntry in SYCL Offload Wrapping.

maksimsab committed:
Merge branch 'main' into public_offload_wrapper

maksimsab committed:
Provide more accurate function name and description.
Also fix failing CI.

jhuber6 commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -754,6 +762,36 @@ bundleOpenMP(ArrayRef<OffloadingImage> Images) {
+
+    for (const OffloadFile &File : OffloadBinaries) {
+      const OffloadBinary &Binary = *File.getBinary();
+      SmallString<0> SerializedImage =
+          OffloadBinary::write(Binary.getOffloadingImage());
```
This is throwing me for a loop a bit and I'm probably missing some context, an `OffloadingImage` is just a C++ format that is used to create the offloading binary. Once the binary has been created why do we need to then go backwards to a C++ struct that contains the same information?

YuriPlyakhin replied:
I think it's time for us to meet and discuss all questions during a call. I'll put together some slides and contact you by email to schedule a call.

jhuber6 reviewed (commented):
Thanks for looking into this and removing at least some of the code duplication. So, let me make sure I'm understanding this correctly. The other offloading targets have a big array of entries that apply to every image embedded in the device. The clang compiler emits these while it's compiling each TU itself.

SYCL on the other hand seems to do this all at the end and generates these arrays as-needed. Images are only needed for registration, so I'm still wondering if there isn't some way to just indicate which entries are to be used for which kernels. If I were doing this on the other targets it would just be a symbol lookup.

Could we at least put these in the same section as the other target's offloading entries? I'm also not entirely sure what the heuristic is for determining which symbols go where.

YuriPlyakhin commented on llvm/lib/Frontend/Offloading/OffloadWrapper.cpp:
```diff
@@ -620,6 +635,409 @@ void createRegisterFatbinFunction(Module &M, GlobalVariable *FatbinDesc,
+      EntriesInits.push_back(GV);
+    }
+
+    auto *Arr = ConstantArray::get(ArrayType::get(EntryTy, EntriesInits.size()),
+                                   EntriesInits);
```
@maksimsab , this code doesn't work:
```
clang-linker-wrapper: llvm-project/llvm/lib/IR/Constants.cpp:1326: static llvm::Constant* llvm::ConstantArray::getImpl(llvm::ArrayType*, llvm::ArrayRef<llvm::Constant*>): Assertion `C->getType() == Ty->getElementType() && "Wrong type in array element initializer"' failed.
```
EntryTy is: `%struct.__tgt_offload_entry = type { i64, i16, i16, i32, ptr, ptr, i64, i64, ptr }`
EntriesInits element type is: `ptr`

I guess more modifications are needed either in `emitOffloadingEntry` or in `initOffloadEntriesPerImage`

maksimsab committed:
Fix initOffloadEntriesPerImage

YuriPlyakhin commented on llvm/lib/Frontend/Offloading/OffloadWrapper.cpp:
```diff
@@ -620,6 +635,409 @@ void createRegisterFatbinFunction(Module &M, GlobalVariable *FatbinDesc,
+      Constant *C = addStringToModule(*LI, "__sycl_offload_entry_name");
+      GlobalVariable *GV =
+          emitOffloadingEntry(M, /*Kind*/ OffloadKind::OFK_SYCL, C,
+                              /*Name*/ "__sycl_offload_entry_name", /*Size*/ 0,
```
@maksimsab , /* Name */ should be the entry name. Later it is used to create `constant string used to look up the symbol in the device.` Please, do the update below. Also you can compare generated structures before and after and you'll see that this change is needed. I verified that with this update our E2E test runs:
```suggestion
                              /*Name*/ *LI, /*Size*/ 0,
```

maksimsab committed:
Simplify handling of SYCL images in clang-linker-wrapper.
Now Offload Wrapper handles the binary blob which contains sycl images.
Also, the issue with duplicated entry names strings is resolved.
SYCLWrappingOptions are renamed to SYCLJITOptions.

maksimsab committed:
Merge branch 'main' into public_offload_wrapper

YuriPlyakhin reviewed (commented):
Please, also fix CI, if it is caused by your changes.

maksimsab committed:
Return LIT testing.
Rollback unittest to resolve CI.
Clean leftovers and unify names.

maksimsab committed:
Remove unnecessary copy of OffloadBinary in Offload Wrapper.

jhuber6 reviewed (approved):
This is generally fine from my side, I'll leave it to the Intel people to understand the specifics.

sarnex commented:
@maksimsab Can you ping me if this is really ready for review? Thanks

sarnex reviewed (approved):
lgtm, all my comments can be addressed in a separate pr or as part of the next iteration of this change. if you are ready otherwise we can merge this before replying to any of my comments

maksimsab committed:
Remove leftovers. Remove unnecessary function. Improve the comment.

maksimsab committed:
Merge branch 'main' into public_offload_wrapper

maksimsab commented on llvm/include/llvm/Frontend/Offloading/OffloadWrapper.h:
```diff
@@ -52,6 +54,23 @@ LLVM_ABI llvm::Error wrapHIPBinary(llvm::Module &M, llvm::ArrayRef<char> Images,
+/// \param Options Data that needs to be encoded for the later use in a runtime.
+LLVM_ABI llvm::Error
+wrapSYCLBinaries(llvm::Module &M, llvm::ArrayRef<char> Buffer,
+                 SYCLJITOptions Options = SYCLJITOptions());
```
Yes, AOT mode is a case when it is too late to pass any options.

maksimsab commented on llvm/lib/Frontend/Offloading/OffloadWrapper.cpp:
```diff
@@ -620,6 +635,410 @@ void createRegisterFatbinFunction(Module &M, GlobalVariable *FatbinDesc,
+  ///   // the entry table
+  ///   __tgt_offload_entry *EntriesBegin;
+  ///   __tgt_offload_entry *EntriesEnd;
+  ///   const char *PropertiesBegin;
+  ///   const char *PropertiesEnd;
```
I needed some "pointer type" here for now either `const char*` or `const void*`. Later, it would be changed to a pointer to appropriate structure.

maksimsab commented on llvm/lib/Frontend/Offloading/OffloadWrapper.cpp:
```diff
@@ -620,6 +635,410 @@ void createRegisterFatbinFunction(Module &M, GlobalVariable *FatbinDesc,
+  GlobalVariable *addGlobalArrayVariable(const Twine &Name,
+                                         ArrayRef<char> Initializer,
+                                         const Twine &Section = "") {
+    auto *Arr = ConstantDataArray::get(M.getContext(), Initializer);
```
In case of name collisions LLVM uses different suffices for the naming.

maksimsab commented on llvm/lib/Frontend/Offloading/OffloadWrapper.cpp:
```diff
@@ -620,6 +635,410 @@ void createRegisterFatbinFunction(Module &M, GlobalVariable *FatbinDesc,
+  GlobalVariable *addGlobalArrayVariable(const Twine &Name,
+                                         ArrayRef<char> Initializer,
+                                         const Twine &Section = "") {
+    auto *Arr = ConstantDataArray::get(M.getContext(), Initializer);
```
Hmm, yes, I see. You are right.

maksimsab commented on llvm/lib/Frontend/Offloading/OffloadWrapper.cpp:
```diff
@@ -660,3 +1079,22 @@ Error offloading::wrapHIPBinary(Module &M, ArrayRef<char> Image,
+  GlobalVariable *Desc = W.createFatbinDesc(OffloadFiles);
+  if (!Desc)
+    return createStringError(inconvertibleErrorCode(),
+                             "No binary descriptors created.");
```
In fact, both options don't follow LLVM's guideline because we don't add any useful piece of information.
